baas-node-state/src/main/scala/com/ing/baker/baas/state/ServiceDiscovery.scala

```diff
@@ -106,39 +94,62 @@ object ServiceDiscovery extends LazyLogging {
-      cacheRecipeListeners <- Stream.eval(Ref.of[IO, Map[RecipeName, List[RecipeListener]]](Map.empty))
-      cacheBakerListeners <- Stream.eval(Ref.of[IO, List[BakerListener]](List.empty))
-      service = new ServiceDiscovery(cacheInteractions, cacheRecipeListeners, cacheBakerListeners)
-      updateServices = fetchServices(namespace, api)
-        .flatMap { currentServices =>
-          List(
-            buildInteractions(currentServices).flatMap(cacheInteractions.set),
-            cacheRecipeListeners.set(buildRecipeListeners(currentServices)),
-            cacheBakerListeners.set(buildBakerListeners(currentServices))
-          ).parSequence
+    def updateServicesWith(
+      currentServices: Ref[IO, List[Service]],
+      cacheInteractions: Ref[IO, List[InteractionInstance]],
+      cacheRecipeListeners: Ref[IO, Map[RecipeName, List[RecipeListener]]],
+      cacheBakerListeners: Ref[IO, List[BakerListener]]
+    )(event: K8SWatchEvent[Service]): IO[Unit] = {
+      for {
+        services <- event._type match {
+          case EventType.ADDED =>
+            currentServices.updateAndGet(event._object :: _)
+          case EventType.DELETED =>
+            currentServices.updateAndGet(_.filterNot(_ == event._object))
+          case EventType.MODIFIED =>
+            currentServices.updateAndGet(_.map(service => if (service.metadata.uid == event._object.metadata.uid) event._object else service))
+          case EventType.ERROR =>
+            IO(logger.error(s"Event type ERROR on service watch for service ${event._object}")) *> currentServices.get
         }
-      updater = Stream.fixedRate(5.seconds).evalMap(_ => updateServices)
-      _ <- Stream.eval(updateServices).concurrently(updater)
-    } yield service
+        _ <- List(
+          buildInteractions(services).flatMap(cacheInteractions.set),
+          cacheRecipeListeners.set(buildRecipeListeners(services)),
+          cacheBakerListeners.set(buildBakerListeners(services))
+        ).parSequence
+      } yield ()
+    }
+
+    val createServiceDiscovery = for {
+      currentServices <- Ref.of[IO, List[Service]](List.empty)
+      cacheInteractions <- Ref.of[IO, List[InteractionInstance]](List.empty)
+      cacheRecipeListeners <- Ref.of[IO, Map[RecipeName, List[RecipeListener]]](Map.empty)
+      cacheBakerListeners <- Ref.of[IO, List[BakerListener]](List.empty)
+      service = new ServiceDiscovery(cacheInteractions, cacheRecipeListeners, cacheBakerListeners)
+      updateServices = updateServicesWith(currentServices, cacheInteractions, cacheRecipeListeners, cacheBakerListeners) _
+      killSwitch <- IO {
+        k8s.watchAllContinuously[Service]()
+          .viaMat(KillSwitches.single)(Keep.right)
+          .toMat(Sink.foreach(updateServices(_).unsafeRunAsyncAndForget()))(Keep.left)
+          .run()
+      }
+    } yield (service, killSwitch)
 
-    stream.compile.resource.lastOrError
+    Resource.make(createServiceDiscovery)(s => IO(s._2.shutdown())).map(_._1)
```

Comment:
a partial function accepting tuples like `case (key, value) => { some code using key and value}  ` looks a bit better than  `_._1` and `_._.2`
at the expense of few extra characters, yes

Reply:
Oh not the tuple discussion again haha, I find these topics a bit opinionated, tbh I find no issues with tuple projections nor pattern matching them first to assign them name in favor of readability. Next time I will do the latter if the team finds issues with tuples, but tbh this is something too tribial to waste our time and energy discussing on it